kernel/aster-nix/src/fs/ramfs/fs.rs

```diff
@@ -728,57 +772,68 @@ impl Inode for RamInode {
         self_dir.remove_entry(idx);
         self_inode.dec_size();
         self_inode.dec_nlinks();
-        drop(self_inode);
-        let mut target_inode = target.0.write();
         target_inode.dec_nlinks();
         target_inode.dec_nlinks();
         Ok(())
     }
 
     fn lookup(&self, name: &str) -> Result<Arc<dyn Inode>> {
-        if self.0.read().metadata.type_ != InodeType::Dir {
-            return_errno_with_message!(Errno::ENOTDIR, "self is not dir");
-        }
-
-        let (_, inode) = self
-            .0
-            .read()
-            .inner
-            .as_direntry()
-            .unwrap()
-            .get_entry(name)
-            .ok_or(Error::new(Errno::ENOENT))?;
+        let inode = self.find(name)?;
         Ok(inode as _)
     }
 
     fn rename(&self, old_name: &str, target: &Arc<dyn Inode>, new_name: &str) -> Result<()> {
-        if self.0.read().metadata.type_ != InodeType::Dir {
-            return_errno_with_message!(Errno::ENOTDIR, "self is not dir");
-        }
-        if !Arc::ptr_eq(&self.fs(), &target.fs()) {
-            return_errno_with_message!(Errno::EXDEV, "not same fs");
-        }
         let target = target
             .downcast_ref::<RamInode>()
             .ok_or(Error::new(Errno::EXDEV))?;
-        if target.0.read().metadata.type_ != InodeType::Dir {
-            return_errno_with_message!(Errno::ENOTDIR, "target is not dir");
-        }
-        if old_name == "." || old_name == ".." {
-            return_errno_with_message!(Errno::EISDIR, "old_name is . or ..");
-        }
-        if new_name == "." || new_name == ".." {
-            return_errno_with_message!(Errno::EISDIR, "new_name is . or ..");
-        }
+
+        let (self_ino, target_ino) = {
+            if !Arc::ptr_eq(&self.fs(), &target.fs()) {
+                return_errno_with_message!(Errno::EXDEV, "not same fs");
+            }
+
+            let (self_type, self_ino) = {
+                let self_inode = self.0.read();
+                (self_inode.metadata.type_, self_inode.metadata.ino as u64)
+            };
+            let (target_type, target_ino) = {
+                let target_inode = target.0.read();
```

Comment:
The ordering of acquiring multiple read locks must be the same as doing multiple write locks. Otherwise, the code lead to deadlock.

Consider the following race conditions where thread 1 acquires lock A, then lock B, while thread 2 acquires lock B, then A.

```
Atomic steps  | Thread 1       | Thread 2
1             | Read lock A    |
2             |                | Write lock B
3             | Read lock B    |
              | (Blocked)      |
4             |                | Write lock A
              |                | (Blocked)
```

The two threads will be blocked forever.